### PR TITLE
Combine ccm data with district info and output to csv

### PR DESCRIPTION
**Pull request by seb231**
Seb: I added some checks to the `add-district-to-dataset-per-user-input` function. Try it and tell me if you find it "safer".
I honestly didn't see the time difference between one function or the other.

### CODE REVIEW

File: src/witan/models/run_models.clj [hunk 1]
```diff
@@ -102,10 +103,26 @@
       (data-csv/write-csv out-file
                           (convert-row-maps-to-vector-of-vectors ordered-data)))))
 
+(defn get-district [gss-code]
+  (let [district-lookup (read-string (slurp (str "resources/default_datasets/local_authorities.edn")))]
```
Comment (Eleonore9) [inline src/witan/models/run_models.clj]: I'm not sure why you need `str` around the filepath. But an alternative would be to use `clojure.edn` like:

``` Clojure
(clojure.edn/read-string (slurp "resources/default_datasets/local_authorities.edn"))
```

File: src/witan/models/run_models.clj [hunk 2]
```diff
@@ -102,10 +103,26 @@
       (data-csv/write-csv out-file
                           (convert-row-maps-to-vector-of-vectors ordered-data)))))
 
+(defn get-district [gss-code]
```
Comment (Eleonore9) [inline src/witan/models/run_models.clj]: I was thinking of def-ing the reading of the edn file so we don't read it each time we do a lookup.
Something like:

``` Clojure
(def local-authorities
    (:gss-code-to-district 
         (clojure.edn/read-string (slurp "resources/default_datasets/local_authorities.edn"))))
```

Comment (Eleonore9) [inline src/witan/models/run_models.clj]: The lookup would then be 

``` Clojure
(get local-authorities (keyword gss-code))
```

We could easily use it to print the district name to the terminal at the start of the programme (after reading the user inputs). That's something Bruce suggested so we're sure we're running the CCM for the right district before seeing the output csv.

File: src/witan/models/run_models.clj [hunk 3]
```diff
@@ -102,10 +103,28 @@
       (data-csv/write-csv out-file
                           (convert-row-maps-to-vector-of-vectors ordered-data)))))
 
+(def local-authorities (read-string (slurp "resources/default_datasets/local_authorities.edn")))
+
+(defn get-district [gss-code]
+  (get-in local-authorities [:gss-code-to-district (keyword gss-code)]))
+
+(defn add-district-to-dataset [data] (wds/add-derived-column data :district [:gss-code] (fn [gss-code] (get-district gss-code))))
```
Comment (Eleonore9) [inline src/witan/models/run_models.clj]: It feels like `add-district-to-dataset` and `add-district-to-dataset2` could be merged into one function.
That would also solve the naming issue.

File: src/witan/models/run_models.clj [hunk 4]
```diff
@@ -102,10 +103,28 @@
...
    Returns a dataset with historical data and projections."
   [inputs gss-code params]
   (core/looping-test (get-datasets inputs gss-code)
                      params))
+
+(defn output-ccm
+  "runs the ccm, adds district information for easier human
+   reading and outputs to a csv"
+  [inputs gss-code params filepath]
+  (prn (str "Preparing projection for " (get-district gss-code) "..."))
```
Comment (Eleonore9) [inline src/witan/models/run_models.clj]: It's just bikeshedding, but I remember Bruce often recommend the use of `format` over `str` (https://clojuredocs.org/clojure.core/format).

File: src/witan/models/run_models.clj [hunk 5]
```diff
@@ -102,10 +103,28 @@
       (data-csv/write-csv out-file
                           (convert-row-maps-to-vector-of-vectors ordered-data)))))
 
+(def local-authorities (read-string (slurp "resources/default_datasets/local_authorities.edn")))
+
+(defn get-district [gss-code]
+  (get-in local-authorities [:gss-code-to-district (keyword gss-code)]))
+
+(defn add-district-to-dataset-per-row [data] (wds/add-derived-column data :district [:gss-code] (fn [gss-code] (get-district gss-code))))
```
Comment (elisehuard) [inline src/witan/models/run_models.clj]: bikeshed: this and next function could be spread over more than 1 line for readability

File: src/witan/models/run_models.clj [hunk 6]
```diff
@@ -102,10 +103,28 @@
...
    Returns a dataset with historical data and projections."
   [inputs gss-code params]
   (core/looping-test (get-datasets inputs gss-code)
                      params))
+
+(defn output-ccm
+  "runs the ccm, adds district information for easier human
+   reading and outputs to a csv"
+  [inputs gss-code params filepath]
+  (println (format "Preparing projection for %s... " (get-district gss-code)))
```
Comment (elisehuard) [inline src/witan/models/run_models.clj:134]: better to use a logging lib (timbre or slf4j) from the start for such things

File: src/witan/models/run_models.clj (hunk 5 above)
Comment (Eleonore9) [inline src/witan/models/run_models.clj]: Good point @elisehuard! That'd help with readability. But I think `cljfmt` will fix it.

**Comment by elisehuard** [comment]
another bikeshed :) it's nice to have an overview of what the commit is achieving in the PR.
Apart from the few bikesheds, lgtm :shipit:

**Comment by seb231** [comment]
Yeah apologies @elisehuard been awhile! should have squashed first too!

**Comment by Eleonore9** [comment]
Let's finish up finding the right `add-district-to-dataset` function and then :shipit:

**Comment by seb231** [comment]
@Eleonore9 that fix you've done for 'add-district-to-dataset-per-user-input' lgtm that's a :shipit: from me

**Comment by jasebell** [comment]
lgtm :shipit:

**Pull request merged**